In [8]:
import os
import datetime 
import pandas as pd 

import openai 
from dotenv import load_dotenv

BASE_DIR = "/".join(os.getcwd().split("/")[: -1])
data_cleaned_dir = "/data/cleaned/yield_curve_historical_rates_MASTER.parquet"

CURRENT_DATE = str(datetime.datetime.today()).split(" ")[0]

load_dotenv()

True

In [9]:
str(datetime.datetime.today()).split(" ")[0]

'2024-05-21'

In [10]:
df = pd.read_parquet(BASE_DIR + data_cleaned_dir)
print(df.shape)

df.head()

(12558, 16)


Date  1 Mo  2 Mo  3 Mo  6 Mo  1 Yr  2 Yr  3 Yr  5 Yr  7 Yr  10 Yr  \
0  2024-05-20  5.50  5.47  5.45  5.43  5.15  4.82  4.62  4.46  4.44   4.44   
1  2024-05-19  5.50  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   
2  2024-05-18  5.50  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   
3  2024-05-17  5.50  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   
4  2024-05-16  5.51  5.47  5.45  5.41  5.13  4.78  4.58  4.40  4.39   4.38   

   20 Yr  30 Yr  2 Yr - 10 Yr  2 Yr - 30 Yr  10 Yr - 30 Yr  
0   4.68   4.58          0.38          0.24          -0.14  
1   4.66   4.56          0.41          0.27          -0.14  
2   4.66   4.56          0.41          0.27          -0.14  
3   4.66   4.56          0.41          0.27          -0.14  
4   4.62   4.52          0.40          0.26          -0.14

In [14]:
df.iloc[: 4]

Date  1 Mo  2 Mo  3 Mo  6 Mo  1 Yr  2 Yr  3 Yr  5 Yr  7 Yr  10 Yr  \
0  2024-05-20   5.5  5.47  5.45  5.43  5.15  4.82  4.62  4.46  4.44   4.44   
1  2024-05-19   5.5  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   
2  2024-05-18   5.5  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   
3  2024-05-17   5.5  5.47  5.46  5.41  5.14  4.83  4.60  4.44  4.43   4.42   

   20 Yr  30 Yr  2 Yr - 10 Yr  2 Yr - 30 Yr  10 Yr - 30 Yr  
0   4.68   4.58          0.38          0.24          -0.14  
1   4.66   4.56          0.41          0.27          -0.14  
2   4.66   4.56          0.41          0.27          -0.14  
3   4.66   4.56          0.41          0.27          -0.14

In [11]:
prompt = """
    **US Treasury Yield Curve Movements On 2024-05-20**

    **Overview:**
    <PROVIDE A BRIEF OVERVIEW OF THE CURRENT STATE OF THE YIELD CURVE AND THE MOST RECENT YIELD CURVE DYNAMICS THAT WOULD BE RELEVANT TO ECONOMISTS AND INVESTORS. COMMENT ON IF THE CURVE STEEPEND OR FLATTENED FOR EXAMPLE.>

    **What it Means**
    1. **Economic Outlook:** <ONE TO THREE SENTENCES ON THE ECONOMIC OUTLOOK BASED ON THE YIELD CURVE>
    
    2. **Monetary Policy:** <ONE TO THREE SENTENCES ON WHAT THE CURRENT YIELD CURVE MIGHT INDICATE FOR MONETARY POLICY>

    3. **Investment Strategies:** <GIVE A BRIEF COMMENT ON WHAT THE CURRENT YIELD CURVE MIGHT INDICATE FOR INVESTORS>
"""

In [15]:
# Create a function to generate insights
def generate_insight(data_summary) -> str:
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are an economic expert on the US Treasury Yield Curve. \
                                           You provide intelligent, objective, non-speculative analyses of recent data.\
                                           You provide these analyses daily after the closing values of the Yield Curve are determined."},
            {"role": "user", "content": f"Provide insights based on the following data in the format below, fill in your analyses in the carrots <>:\
                                          The response template you need to use: {prompt}. \
                                          The summary data: {data_summary}."}
        ]
    )
    return response['choices'][0]['message']['content'].strip()

# Get insights for the DataFrame summary
insights = generate_insight(data_summary=df.iloc[: 50])
print(insights)

**US Treasury Yield Curve Movements On 2024-05-20**

**Overview:**
The US Treasury yield curve experienced a slight flattening on May 20, with short-term yields holding steady and longer-term yields slightly decreasing.

**What it Means**
1. **Economic Outlook:** The flattening of the yield curve suggests that investors may have slightly reduced expectations for future economic growth.
    
2. **Monetary Policy:** The stability in short-term yields indicates a lack of immediate concerns about inflation or the need for aggressive monetary tightening by the Federal Reserve.
    
3. **Investment Strategies:** Investors may view longer-term bonds as slightly more attractive in the current environment of lower long-term yields, potentially reallocating their portfolios towards these securities.


**US Treasury Yield Curve Movements On 2024-05-20**

**Overview:**
The yield curve on May 20, 2024, indicates a gradual decrease in yields from the short-end (1-month: 5.50%) to the long-end (30-year: 4.58%). This indicates a downward sloping or "inverted" yield curve. The yield differential between 2-year and 10-year (0.38) is slightly positive, while the difference between 2-year and 30-year treasuries shows a small inverse (0.24). The yield differential between the 10-year and 30-year is negative (-0.14), representing a flatter yield curve.

**What it Means**
1. **Economic Outlook:** The inverted yield curve generally signifies a pessimistic economic outlook, as investors are demanding higher yields for shorter-term securities compared to longer-term ones, typically an indicator of expected economic slowdown or recession.

2. **Monetary Policy:** In terms of monetary policy, the current yield curve might indicate a tightening monetary environment. The Federal Reserve might consider easing its monetary controls (like interest rate reductions) to prop up the economy and rectify the inversion.

3. **Investment Strategies:** For investors, this type of yield curve could suggest moving investments towards shorter-term securities to capitalize on their higher yields compared to long-term bonds. However, they must also consider their risk appetite and long term investment objectives as investing in shorter-term securities could entail more frequent reinvestment risks.